File: src/test/java/ar/com/flow/bankaccount/withdrawal/WithdrawalTest.java
package ar.com.flow.bankaccount.withdrawal;

import ar.com.flow.Customer;
import ar.com.flow.bankaccount.Balance;
import ar.com.flow.bankaccount.TestObjects;
import ar.com.flow.money.InsufficientFundsException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static ar.com.flow.money.TestObjects.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class WithdrawalTest {
    private Customer francisco;

    @BeforeEach
    void setUp() {
        francisco = Customer.named("francisco");
    }

    @Test
    void givenSavingsAccountWith100USDBalanceWhenExecute10USDWithdrawalThenItShouldCalculateNewBalance() {
        var account = TestObjects.createSavingsAccountFor(francisco, dollars100);

        var transactionRecord = Withdrawal.from(account).amount(dollars10).execute();

        assertEquals(Balance.negative(dollars10), transactionRecord.getBalance());
    }

    @Test
    void givenSavingsAccountWith100USDBalanceWhenCreate200USDWithdrawalThenItShouldBeRejected() {
        var account = TestObjects.createSavingsAccountFor(francisco, dollars100);

        assertThrows(InsufficientFundsException.class, () -> Withdrawal.from(account).amount(dollars200).execute());
    }
}

File: src/test/kotlin/ar/com/flow/bankaccount/usecases/WithdrawalTest.kt
package ar.com.flow.bankaccount.usecases

import ar.com.flow.Customer
import ar.com.flow.bankaccount.adapters.out.persistence.memory.InMemoryBankAccounts
import ar.com.flow.bankaccount.adapters.out.persistence.memory.InMemoryCustomers
import ar.com.flow.bankaccount.domain.Balance
import ar.com.flow.bankaccount.domain.Currency
import ar.com.flow.bankaccount.ports.out.BankAccounts
import ar.com.flow.bankaccount.ports.out.Customers
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class WithdrawalTest {
    private lateinit var customers: Customers
    private lateinit var bankAccounts: BankAccounts

    @BeforeEach
    internal fun setUp() {
        customers = InMemoryCustomers()
        bankAccounts = InMemoryBankAccounts()
    }

    @Test
    internal fun withdraw() {
        val juanPerez = customers.save(Customer("Juan Perez"))
        bankAccounts.create(juanPerez, currency = Currency.ARS)

        val deposit = Deposit(customers, bankAccounts)
        deposit.execute(customerName = "Juan Perez", currencyCode = Currency.ARS.code, amountToDeposit = 100)

        val withdrawal = Withdrawal(customers, bankAccounts)
        withdrawal.execute(customerName = "Juan Perez", Currency.ARS.code, amountToWithdraw = 100)

        val savingsAccount = bankAccounts.ownedBy(juanPerez, Currency.ARS).first()
        assertThat(savingsAccount.balance).isEqualTo(Balance(Currency.ARS, 0))
    }
}